src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,113 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elementData;
+    private int size;
+
+    public ArrayList() {
+        elementData = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        growIfArrayFull();
+        elementData[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexOutOfBoundsInclusive(index);
+        growIfArrayFull();
+        if (index != size) {
+            System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        }
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexOutOfBoundsExclusive(index);
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndexOutOfBoundsExclusive(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexOutOfBoundsExclusive(index);
+        T removedElement = elementData[index];
+        if (index != --size) {
+            System.arraycopy(elementData, index + 1, elementData, index, size - index);
+        }
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T removedElement = null;
+        for (int i = 0; i < size; i++) {
+            if (elementData[i] == element
+                    || elementData[i] != null && elementData[i].equals(element)) {
+                removedElement = remove(i);
+                return removedElement;
+            }
+        }
+        checkNoSuchElementException(removedElement);
+        return removedElement;
```

Comment:
```suggestion
        throw new NoSuchElementException("Value does not exist! " + element);
```
we will reach this point only if we didn't find equal element